src/libexpr/c/nix_api_value.h

```diff
@@ -0,0 +1,386 @@
+#ifndef NIX_API_VALUE_H
+#define NIX_API_VALUE_H
+
+/** @addtogroup libexpr
+ * @{
+ */
+/** @file
+ * @brief libexpr C bindings dealing with values
+ */
+
+#include "nix_api_util.h"
+#include "stdbool.h"
+#include "stddef.h"
+#include "stdint.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+// cffi start
+
+// Type definitions
+typedef enum {
+    NIX_TYPE_THUNK,
```

Comment:
With the Nix team we've previously discussed that perhaps THUNK should be removed from the interface.

All the accessors, including `nix_get_type` could then force the value if still needed.

A function `nix_is_thunk` could now or eventually be added to recover the ability for the bindings to observe whether a value is still a thunk or not (which would be an advanced feature).

As a side effect, many more functions would need a `State`, which is a good thing because it buys more flexibility for the bindings to endure changes in the evaluator. We don't want to be stuck in a situation where our callers don't give us a `State` because previous versions of Nix did not require it.

(I've experienced this with changes to `StorePath`-related function and method signatures in the context of third party bindings. `StorePath` is to `Store` as `Value` is to `State`, somewhat.)

Reply:
The reason I kept thunks is for implementing primops. You don't want them to be always strict in the arguments.

Reply:
The issue with every accessor forcing thunks, in my opinion, is that every accessor can now throw any error, which is a lot more annoying to deal with, in my opinion. I'd probably be okay with a `get_type` that forces, but I'm not sure if `nix_get_int` being able to throw evaluation errors is very useful or desired when writing bindings (you'll have to call `nix_get_type` first anyways, always)

Reply:
> > All the accessors, including `nix_get_type` could then force the value if still needed.

> every accessor can now throw any error

I think I was wrong to ask for all the accessors to get this treatment.

If the code forces values step by step, as I believe it should, it benefits from a forcing `nix_get_type`, as explained below, but if the code does a `deepSeq`, it seems that it can mostly do away with error handling, which could be considered a feature.

Nonetheless a forcing `get_type` seems like a nice interface, so perhaps we should have it in addition; e.g. `nix_force_type`? The existence of the THUNK constant shouldn't be a problem, because forward compatible clients should have a suitable `default:` clause anyway.

<details><summary>Now-redundant explanation, non-deepSeq case only</summary>

`nix_get_int` can definitely fail if used incorrectly, on a non-int value. That's fundamental, given your freedom to call it before any forcing or runtime type checking is done. It currently has an `assert`, indicating that such usage is a programming error, and not a domain error.
In the current situation, it must be preceded by a forcing call, to satisfy the assertion.
In the new situation, this is still the case, but the forcing can happen through `nix_get_type`.

</details>

**Conclusion**

Keep THUNK, and perhaps add `nix_force_type` for convenience, to do both the forcing and type returning, which is a very common combination.